ports/espressif/common-hal/sdioio/SDCard.c

```diff
@@ -0,0 +1,223 @@
+// This file is part of the CircuitPython project: https://circuitpython.org
+//
+// SPDX-FileCopyrightText: Copyright (c) 2024 Jacob Rigby
+//
+// SPDX-License-Identifier: MIT
+
+#include <stdbool.h>
+#include "esp_err.h"
+
+#include "driver/sdmmc_host.h"
+#include "ports/espressif/esp-idf/components/sdmmc/include/sdmmc_cmd.h"
+
+#include "shared-bindings/microcontroller/Pin.h"
+#include "shared-bindings/util.h"
+#include "shared-bindings/sdioio/SDCard.h"
+#include "py/mperrno.h"
+#include "py/runtime.h"
+
+#include "esp_log.h"
+static const char *TAG = "SDCard.c";
+
+static bool sdio_host_init = false;
+
+static void common_hal_sdioio_sdcard_check_for_deinit(sdioio_sdcard_obj_t *self) {
+    if (common_hal_sdioio_sdcard_deinited(self)) {
+        raise_deinited_error();
+    }
+}
+
+static int check_pins(const mcu_pin_obj_t *clock, const mcu_pin_obj_t *command, const uint8_t num_data, const mcu_pin_obj_t **data) {
+    if (CONFIG_SOC_SDMMC_USE_GPIO_MATRIX) {
+        // ESP32-S3 and P4 can use any pin for any SDMMC func in either slot
+        // Default to SLOT_1 for SD cards
+        ESP_LOGI(TAG, "Using chip with CONFIG_SOC_SDMMC_USE_GPIO_MATRIX");
+        return SDMMC_HOST_SLOT_1;
+    }
+    if (command->number == GPIO_NUM_11 && clock->number == GPIO_NUM_6 && data[0]->number == GPIO_NUM_7) {
+        // Might be slot 0
+        if (num_data == 1 || (num_data == 4 && data[1]->number == GPIO_NUM_8 && data[2]->number == GPIO_NUM_9 && data[3]->number == GPIO_NUM_10)) {
+            return SDMMC_HOST_SLOT_0;
+        }
+    } else if (command->number == GPIO_NUM_15 && clock->number == GPIO_NUM_14 && data[0]->number == 2) {
+        // Might be slot 1
+        if (num_data == 1 || (num_data == 4 && data[1]->number == GPIO_NUM_4 && data[2]->number == GPIO_NUM_12 && data[3]->number == GPIO_NUM_13)) {
+            return SDMMC_HOST_SLOT_1;
+        }
+    }
+    return -1;
+}
+
+void common_hal_sdioio_sdcard_construct(sdioio_sdcard_obj_t *self,
+    const mcu_pin_obj_t *clock, const mcu_pin_obj_t *command,
+    uint8_t num_data, const mcu_pin_obj_t **data, uint32_t frequency) {
+    if (num_data != 4 && num_data != 1) {
+        mp_raise_ValueError(MP_ERROR_TEXT("SDIO: cards have either 1 or 4 data lines"));
```

Comment:
There is probably a message we can reuse here too.

Reply:
I looked briefly but not carefully, I'll look again :grin:

Reply:
Thanks, done.